arch/riscv/adler32_rvv.c

```diff
@@ -12,23 +12,25 @@
 #include "../../zbuild.h"
 #include "../../adler32_p.h"
 
-Z_INTERNAL uint32_t adler32_rvv(uint32_t adler, const uint8_t *buf, size_t len) {
+static inline uint32_t adler32_rvv_impl(uint32_t adler, uint8_t* restrict dst, const uint8_t *src, size_t len, int COPY) {
```

Comment:
Is there an opportunity to optimize `adler32_avx2` and `adler32_avx512` implementations to use `int COPY`?

Reply:
It's a little less explicit than the templatized variant but assuming the compiler can do the most basic of keyhole optimizations, it'd probably compile to the same result (without the preprocessor mess).

Reply:
Let me give it a whirl.  I don't have a strong feeling one way or the other as long as it works but others might have stronger opinions about a compiler's ability to hoist the variable out entirely.

Reply:
Yeah I don't see anything I'd consider a serious regression, here.  It did slightly reorder the instructions but it definitely doesn't branch:
### Old
``` 
13.24 │ a0:   vmovdqu       (%r8,%rax,1),%ymm0
  0.21 │       vmovdqa       %ymm3,%ymm2
 13.19 │       vpsadbw       %ymm8,%ymm0,%ymm1
  0.38 │       vmovdqu       %ymm0,(%rsi,%rax,1)
  9.13 │       vpmaddubsw    %ymm7,%ymm0,%ymm0
  8.22 │       vpmaddwd      %ymm6,%ymm0,%ymm0
 10.79 │       vpaddd        %ymm1,%ymm3,%ymm1
  0.27 │       vpaddd        %ymm5,%ymm2,%ymm2
 18.89 │       vpaddd        %ymm4,%ymm0,%ymm0
  0.55 │       add           $0x20,%rax
  8.07 │       vmovdqa       %ymm1,%ymm3
  0.26 │       vmovdqa       %ymm2,%ymm5
 13.21 │       vmovdqa       %ymm0,%ymm4
       │       cmp           %rdx,%rax
  0.34 │     ↑ jne           a0
```

### New
```
 11.61 │ a0:   vmovdqu       (%r9,%rdx,1),%ymm0
  0.27 │       vmovdqa       %ymm3,%ymm2
 13.64 │       vpsadbw       %ymm8,%ymm0,%ymm1
  0.32 │       vmovdqu       %ymm0,(%r8,%rdx,1)
 10.22 │       vpmaddubsw    %ymm7,%ymm0,%ymm0
  4.11 │       vpmaddwd      %ymm6,%ymm0,%ymm0
 11.88 │       vpaddd        %ymm1,%ymm3,%ymm1
  0.38 │       vpaddd        %ymm5,%ymm2,%ymm2
 17.76 │       vpaddd        %ymm4,%ymm0,%ymm0
  0.46 │       add           $0x20,%rdx
 10.58 │       vmovdqa       %ymm1,%ymm3
  0.30 │       vmovdqa       %ymm2,%ymm5
 14.54 │       vmovdqa       %ymm0,%ymm4
       │       cmp           %rax,%rdx
  0.25 │     ↑ jne           a0
```

Reply:
GCC and Clang support `__attribute__((always_inline))` if you really want to nail it down, but that's a thing you'd want to abstract to cmake to replace it with a portable macro.

Or go nuclear.  Write the whole thing in C++ and template it.

Reply:
Rewriting C code in C++ is recipe for disaster... C++ might allow a lot fancier code, but it comes with quite a big overhead. It's also common to not install C++ compiler or it's dependencies by default, but still have C compiler. This makes using for example clang really hard in some operating systems as one has to install support libraries for g++ first before trying to use clang, as clang is essentially a C++ compiler just like VC++ on Windows. Some code that is valid with C compiler isn't valid with C++ compiler.

During the 40 years I have been a developer, I've seen a lot of projects just rename *.c files to *.cpp or *.cxx just to pretend the code compiles better and still works as expected. There is also a running joke about a lot of C++ headers just being one line that includes the equivalent C header just to conform the naming guidelines of the C++ standard.